src/interpreter/interpreter.kt

```diff
@@ -0,0 +1,195 @@
+package interpreter
+
+import core.*
+import standard_library.stdLib
+import java.lang.Math.pow
+
+fun PreOp.eval(arg : Value) = when(this) {
+    PreOp.plus -> arg as? VInt ?: throw IncorrectTypeException(wrongVal = arg)
+    PreOp.negate -> if(arg is VInt) VInt(-arg.value) else throw IncorrectTypeException(wrongVal = arg)
+    PreOp.not -> if(arg is VBool) VBool(!arg.value) else throw IncorrectTypeException(wrongVal = arg)
+}
+
+fun InOp.eval(lhs : Value, rhs : Value) : Value {
+    val iLeft by lazy { (lhs as? VInt ?: throw IncorrectTypeException(wrongVal = lhs)).value }
```

Comment:
isnt the by operator lazy already? why is lazy {} required

Reply:
No, see https://kotlinlang.org/docs/reference/delegated-properties.html

Reply:
I'm not sure if it's better to do it this way though. It's kind of wasteful to make 6 thunks when we're only going to need 2 of them (and if we e.g. add a cons operator then that's 8 thunks, and 10 if we allow operators to act on functions).

Reply:
It's just gross to have it expanded in every case

Reply:
I just checked the implementation and apparently the default lazy delegate handles thread safety. Do we care about thread safety? Because otherwise I'll replace those with `lazy(LazyThreadSafetyMode.NONE) { ... }`

Reply:
I thoroughly misunderstood the by operator, thanks to the docs I understand it now and it's pretty cool!

It is wasteful but imo its fine cause it's much much more aesthetically pleasing how it is.

I don't know about thread safety we can talk about it later.